src/VisualStudio/Core/Def/Workspace/VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory.cs

```diff
@@ -0,0 +1,105 @@
+    internal class VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory : IWorkspaceServiceFactory, IVsSolutionEvents
+    {
+        /// <summary>
+        /// The collector that's used to collect all the telemetry for operations within <see cref="VisualStudioWorkspace"/>. We'll report this
+        /// when the solution is closed, so the telemetry is linked to that.
+        /// </summary>
+        private readonly SourceGeneratorTelemetryCollectorWorkspaceService _visualStudioWorkspaceInstance = new SourceGeneratorTelemetryCollectorWorkspaceService();
+
+        /// <summary>
+        /// The collector used to collect telemetry for any other workspaces that might be created; we'll report this at the end of the session since nothing here is necessarily
+        /// linked to a specific solution. The expectation is this may be empty for many/most sessions, but we don't want a hole in our reporting and discover that the hard way.
+        /// </summary>
+        private readonly SourceGeneratorTelemetryCollectorWorkspaceService _otherWorkspacesInstance = new SourceGeneratorTelemetryCollectorWorkspaceService();
+
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsyncServiceProvider _serviceProvider;
+        private volatile int _subscribedToSolutionEvents;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
+        {
+            _threadingContext = threadingContext;
+            _serviceProvider = (IAsyncServiceProvider)serviceProvider;
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            // We will record all generators for the main workspace in one bucket, and any other generators running in other
+            // workspaces (interactive, for example) will be put in a different bucket. This allows us to report the telemetry
+            // from the primary workspace on solution closed, while not letting the unrelated runs pollute those numbers.
+            if (workspaceServices.Workspace is VisualStudioWorkspace)
+            {
+                EnsureSubscribedToSolutionEvents();
+                return _visualStudioWorkspaceInstance;
+            }
+            else
+            {
+                return _otherWorkspacesInstance;
+            }
+        }
+
+        private void EnsureSubscribedToSolutionEvents()
+        {
+            if (Interlocked.CompareExchange(ref _subscribedToSolutionEvents, 1, 0) == 0)
+            {
+                Task.Run(async () =>
+                {
+                    var shellService = await _serviceProvider.GetServiceAsync<SVsSolution, IVsSolution>(_threadingContext.JoinableTaskFactory).ConfigureAwait(true);
```

Comment:
CA(true) intentional?  not sure what purpose that woudl serve?

Reply:
It avoids one extra thread schedule, at least theoretically. If this runs on the UI thread and the GetServiceAsync() happens asynchronously, it means we won't end up scheduling work to the thread pool whose only job is to schedule back to the UI thread. We'd just schedule there immediately and the SwitchToMainThread() then completes immediately.

Here it's probably not necessary since that GetServiceAsync() is going to complete synchronously, but it's a general habit if the next line is just going to be a SwitchToMainThread().